lib/network.go

```diff
@@ -2119,11 +2149,25 @@ func (msg *MsgDeSoHeader) EncodeHeaderVersion2(preSignature bool) ([]byte, error
 	}
 	retBytes = append(retBytes, encodedValidatorsTimeoutAggregateQC...)
 
+	// ProposerVotePartialSignature: we encode the signature if it's present and the preSignature
+	// flag is set to false. Otherwise, we encode an empty byte array as a placeholder. Note, any
+	// byte encoded output from this function must be decodable by the DecodeHeaderVersion2
+	// function.
+	proposerSignatureBytes := []byte{}
+	if !preSignature {
+		// If the the preSignature flag is set to false, then the caller intends to encode the signature.
+		// The signature must be non-nil.
+		if msg.ProposerVotePartialSignature == nil {
+			return nil, fmt.Errorf("EncodeHeaderVersion2: ProposerVotePartialSignature must be non-nil")
+		}
+		proposerSignatureBytes = msg.ProposerVotePartialSignature.ToBytes()
+	}
+	retBytes = append(retBytes, EncodeByteArray(proposerSignatureBytes)...)
```

Comment:
I think this append should only happen inside the `if !preSignature {}` condition. Otherwise, it will append `[]byte{0}` which seems problematic. The `[]byte{0}` is the result of encoding array lenght in `EncodeToBytes`.

To see why this could be problematic, let's say that `X := EncodeHeaderVersion2(true)`. Now say `y := EncodeByteArray(proposerSignatureBytes)`. If you look at `X[: len(X) - len(y)]` I think it's not equal to `EncodeHeaderVersion2(false)`, which is what we supposedly want, because that's what we will be signing. Unless I'm misunderstanding how the preSignature flag is meant to work.

Reply:
We need the append outside of the `if !preSignature {}` so that the DecodeHeaderVersion2 function can decode all possible outputs produced by the EncodeHeaderVersion2 function. Without it, the byte encoding/decoding won't be bijective.

You are correct that it's always the case that `len(EncodeHeaderVersion2(false)) != len(EncodeHeaderVersion2(true)) + len(proposerSignatureBytes)`. This is a result of using variable length ints to store the signature length. IMO, the above isn't problematic though.

The `preSignature` flag usage gives us the critical properties we want:
1. It gives us a deterministic method to encode the header to bytes excluding the signature. The placeholder `[]byte{0}` will always be encoded when `preSignature=true`. This is straightforward to decode if needed.
2. It give us a deterministic method to encode the header including the signature. In this case the end of the encoded bytes hare `[]byte{len(signature)} ... proposerSignatureBytes`. This is also straightforward to decode.
3. Given a header with the signature field populated, it gives us a way to verify the signature field as follows:
  i. Compute `preSignatureBytes = header.ToBytes(preSignature=true)`
  ii. Compute `headerHash = Hash(preSignatureBytes)`
  iii. Check `VerifyBLSSignature(header.ProposerVotingPublicKey, header.ProposerVotePartialSignature, headerHash)`


Can you elaborate on the risks you see with the above?

Reply:
Looks like you got this covered. As long as we're consistent that header byte encoding for hashes always has the extra fictitious []byte{0} at the end, then we're good. I guess "preSignature" is slightly misleading. What it really means is "nill signature", which means an encoding that will not be a slice of the encoding "with signature" included (they differ on that last []byte{0}). Since this is a convention we've used in other place, let's go forward with this.